Modify Rule S5883(C#): Education Framework (APPSEC-61) (#1143)
SonarSource/rspec: rules/S5883/csharp/how-to-fix-it/dotnet.adoc

=== How to fix it in .NET

The following code uses the `find` command and expects the user to enter the
name of a file to find on the system.

It is vulnerable to arguments injection because untrusted data is inserted
directly into the arguments of a process call without sanitization. +
The application assumes that incoming data always consists of a specific range
of characters and ignores that some characters might force the `find` command

> to start a shell.

Author: ```suggestion
to forcefully remove files.
```

> to forcefully remove files in `/some/folder`.

Author: ```suggestion
to remove files in `/some/folder`.
```

> of characters and ignores that some characters might force the `find` command
> to remove files in `/some/folder`.
> 
> In this particular case, an attacker may exploit the query with the following string:

Author: ```suggestion
of characters and ignores that some characters might force the `find` command
to start a shell.

In this particular case, an attacker may remove files in `/some/folder` with the following string:
```

----

> * -exec rm {} ;

Author: ```suggestion
'*' -exec echo rm -rf {} \; 
```

> -exec /bin/sh \; -quit

Author: ah yeah I like the rm one, it's clearer

----

[cols="a"]
|===
h| Non-compliant code example
|
[source,csharp]
----
public class ExampleController : Controller
{
    public void Run(string args)
    {
        Process p             = new Process();
        p.StartInfo.FileName  = "/usr/bin/find";
        p.StartInfo.Arguments = "/some/folder -iname " + args; // Noncompliant
        p.Start();
    }
}
----
h| Compliant solution
|
[source,csharp]
----
public class ExampleController : Controller
{
    public void Run(string args)
    {
        Process p            = new Process();
        p.StartInfo.FileName = "/usr/bin/find";
        p.StartInfo.ArgumentList.Add("/some/folder");
        p.StartInfo.ArgumentList.Add("-iname");
        p.StartInfo.ArgumentList.Add(arg);
        p.Start();
    }
}
----
|===

=== How does this work?

include::../../common/fix/introduction.adoc[]

:sanitizationLib: ArgumentList.Add()
include::../../common/fix/sanitize-meta-characters.adoc[]

Here `ArgumentList` takes care of escaping the passed arguments and internally

> creates a single string given to the operating system when `Process.Start()` is

Reviewer: ```suggestion
creates a single string given to the operating system when `System.Diagnostics.Process.Start()` is
```

The following code uses the `find` command and expects the user to enter the
name of a file to find on the system.

It is vulnerable to arguments injection because untrusted data is inserted
directly into the arguments of a process call without sanitization. +
The application assumes that incoming data always consists of a specific range
of characters and ignores that some characters might force the `find` command
to start a shell.

> In this particular case, an attacker may exploit the query with the following string:

Reviewer: Maybe you can mention that the payload will remove all file for the application root directory?

Author: yup, my bad !
